src/terms/kinetic.jl

```diff
@@ -19,18 +19,27 @@ struct TermKinetic <: Term
     kinetic_energies::Vector{<:AbstractVector}
 end
 function TermKinetic(basis::PlaneWaveBasis{T}, scaling_factor, blowup) where {T}
-    kinetic_energies = [[T(scaling_factor) * sum(abs2, Gk)/2 * blowup(norm(Gk), basis.Ecut)
-                         for Gk in Gplusk_vectors_cart(basis, kpt)] for kpt in basis.kpoints]
+
+    kinetic_energies = [kinetic_energy(Gplusk_vectors_cart(basis, kpt), scaling_factor,
+                                        blowup, basis.Ecut, T)
+                        for kpt in basis.kpoints]
     TermKinetic(T(scaling_factor), kinetic_energies)
 end
 
+function kinetic_energy(q, scaling_factor, blowup, Ecut, ::Type{T}) where {T}
```

Comment:
Get rid of the last argument and infer from the eltype of q, also I would reorder the arguments to
`(blowup, scaling_factor, Ecut, q)`. That allows you to also have a version `kinetic_energy(kin::Kinetic, Ecut, q)`, which simplifies the preconditioner.